In [1]:
import numpy as np
import mnist
from keras.models import Sequential
from keras.layers import Dense
from keras.utils import to_categorical

Using TensorFlow backend.


In [2]:
train_images = mnist.train_images()
train_labels = mnist.train_labels()
test_images = mnist.test_images()
test_labels = mnist.test_labels()

In [3]:
train_images.shape

(60000, 28, 28)

In [4]:
# Normalize the images.
train_images = (train_images / 255) - 0.5
test_images = (test_images / 255) - 0.5

# Flatten the images.
train_images = train_images.reshape((-1, 784))
test_images = test_images.reshape((-1, 784))

In [5]:
# Build the model.
model = Sequential([
  Dense(64, activation='relu', input_shape=(784,)),
  Dense(64, activation='relu'),
  Dense(10, activation='softmax'),
])

# Compile the model.
model.compile(
  optimizer='adam',
  loss='categorical_crossentropy',
  metrics=['accuracy'],
)

# Train the model.
model.fit(
  train_images,
  to_categorical(train_labels),
  epochs=5,
  batch_size=32,
)

# Evaluate the model.
model.evaluate(
  test_images,
  to_categorical(test_labels)
)


Epoch 1/5
60000/60000 [==============================] - 8s 132us/step - loss: 0.3760 - accuracy: 0.8876
Epoch 2/5
60000/60000 [==============================] - 7s 120us/step - loss: 0.1913 - accuracy: 0.9420
Epoch 3/5
60000/60000 [==============================] - 7s 122us/step - loss: 0.1476 - accuracy: 0.9547
Epoch 4/5
60000/60000 [==============================] - 8s 129us/step - loss: 0.1234 - accuracy: 0.9610
Epoch 5/5
10000/10000 [==============================] - 0s 20us/step


[0.11804294485226273, 0.9634000062942505]

Попробуем увеличить кол-во эпох.

In [6]:
# Build the model.
model = Sequential([
  Dense(64, activation='relu', input_shape=(784,)),
  Dense(64, activation='relu'),
  Dense(10, activation='softmax'),
])

# Compile the model.
model.compile(
  optimizer='adam',
  loss='categorical_crossentropy',
  metrics=['accuracy'],
)

# Train the model.
model.fit(
  train_images,
  to_categorical(train_labels),
  epochs=7,
  batch_size=32,
)

# Evaluate the model.
model.evaluate(
  test_images,
  to_categorical(test_labels)
)

Epoch 1/7
60000/60000 [==============================] - 8s 138us/step - loss: 0.3631 - accuracy: 0.8908
Epoch 2/7
60000/60000 [==============================] - 7s 124us/step - loss: 0.1902 - accuracy: 0.9427
Epoch 3/7
60000/60000 [==============================] - 8s 127us/step - loss: 0.1448 - accuracy: 0.9554
Epoch 4/7
60000/60000 [==============================] - 9s 154us/step - loss: 0.1247 - accuracy: 0.9614
Epoch 5/7
60000/60000 [==============================] - 8s 132us/step - loss: 0.1106 - accuracy: 0.9648
Epoch 6/7
60000/60000 [==============================] - 9s 144us/step - loss: 0.0967 - accuracy: 0.9694
Epoch 7/7
10000/10000 [==============================] - 0s 25us/step


[0.13241413424229248, 0.9591000080108643]

Как видно, точность упала - модель слегка переобучилась.

Теперь попробуем увеличить кол-во нейронов в каждом слое

In [11]:
# Build the model.
model = Sequential([
  Dense(74, activation='relu', input_shape=(784,)),
  Dense(74, activation='relu'),
  Dense(10, activation='softmax'),
])

# Compile the model.
model.compile(
  optimizer='adam',
  loss='categorical_crossentropy',
  metrics=['accuracy'],
)

# Train the model.
model.fit(
  train_images,
  to_categorical(train_labels),
  epochs=5,
  batch_size=32,
)

# Evaluate the model.
model.evaluate(
  test_images,
  to_categorical(test_labels)
)

Epoch 1/5
60000/60000 [==============================] - 3s 49us/step - loss: 0.3517 - accuracy: 0.8949
Epoch 2/5
60000/60000 [==============================] - 3s 48us/step - loss: 0.1814 - accuracy: 0.9457
Epoch 3/5
60000/60000 [==============================] - 3s 44us/step - loss: 0.1380 - accuracy: 0.9578
Epoch 4/5
60000/60000 [==============================] - 3s 43us/step - loss: 0.1144 - accuracy: 0.9647
Epoch 5/5
10000/10000 [==============================] - 0s 23us/step


[0.10171364394016563, 0.9666000008583069]

Точность немного возрасла.

Теперь попробуем изменить функцию активации у некоторых слоев.

In [12]:
# Build the model.
model = Sequential([
  Dense(64, activation='sigmoid', input_shape=(784,)),
  Dense(64, activation='sigmoid'),
  Dense(10, activation='softmax'),
])

# Compile the model.
model.compile(
  optimizer='adam',
  loss='categorical_crossentropy',
  metrics=['accuracy'],
)

# Train the model.
model.fit(
  train_images,
  to_categorical(train_labels),
  epochs=5,
  batch_size=32,
)

# Evaluate the model.
model.evaluate(
  test_images,
  to_categorical(test_labels)
)

Epoch 1/5
60000/60000 [==============================] - 9s 142us/step - loss: 0.5652 - accuracy: 0.8561
Epoch 2/5
60000/60000 [==============================] - 8s 137us/step - loss: 0.2291 - accuracy: 0.9325
Epoch 3/5
60000/60000 [==============================] - 9s 153us/step - loss: 0.1703 - accuracy: 0.9497
Epoch 4/5
60000/60000 [==============================] - 9s 148us/step - loss: 0.1339 - accuracy: 0.9605
Epoch 5/5
10000/10000 [==============================] - 0s 25us/step


[0.13725836685821413, 0.9575999975204468]

Сигмоида хуже справляется с задачей нежели relu.

Попробуем добавить еще один слой.

In [13]:
# Build the model.
model = Sequential([
  Dense(64, activation='relu', input_shape=(784,)),
  Dense(64, activation='relu'),
  Dense(64, activation='relu'),
  Dense(10, activation='softmax'),
])

# Compile the model.
model.compile(
  optimizer='adam',
  loss='categorical_crossentropy',
  metrics=['accuracy'],
)

# Train the model.
model.fit(
  train_images,
  to_categorical(train_labels),
  epochs=5,
  batch_size=32,
)

# Evaluate the model.
model.evaluate(
  test_images,
  to_categorical(test_labels)
)

Epoch 1/5
60000/60000 [==============================] - 8s 130us/step - loss: 0.3448 - accuracy: 0.8928
Epoch 2/5
60000/60000 [==============================] - 7s 122us/step - loss: 0.1766 - accuracy: 0.9452
Epoch 3/5
60000/60000 [==============================] - 8s 126us/step - loss: 0.1415 - accuracy: 0.9553
Epoch 4/5
60000/60000 [==============================] - 8s 127us/step - loss: 0.1203 - accuracy: 0.9620
Epoch 5/5
10000/10000 [==============================] - 1s 133us/step


[0.10592387998532504, 0.9656000137329102]

Точность модели немного возрасла по сравнению с исходной.

Теперь попробуем добваить нейронов и новый слой.

In [16]:
# Build the model.
model = Sequential([
  Dense(80, activation='relu', input_shape=(784,)),
  Dense(80, activation='relu'),
  Dense(80, activation='relu'),
  Dense(10, activation='softmax'),
])

# Compile the model.
model.compile(
  optimizer='adam',
  loss='categorical_crossentropy',
  metrics=['accuracy'],
)

# Train the model.
model.fit(
  train_images,
  to_categorical(train_labels),
  epochs=5,
  batch_size=32,
)

# Evaluate the model.
model.evaluate(
  test_images,
  to_categorical(test_labels)
)

Epoch 1/5
60000/60000 [==============================] - 3s 57us/step - loss: 0.3337 - accuracy: 0.8970
Epoch 2/5
60000/60000 [==============================] - 3s 51us/step - loss: 0.1652 - accuracy: 0.9491
Epoch 3/5
60000/60000 [==============================] - 3s 51us/step - loss: 0.1313 - accuracy: 0.9592
Epoch 4/5
60000/60000 [==============================] - 3s 51us/step - loss: 0.1102 - accuracy: 0.9650
Epoch 5/5
10000/10000 [==============================] - 0s 27us/step


[0.10811293295091018, 0.9653000235557556]

Результат получился чуть хуже, попробуем добавить еще нейронов.

In [18]:
# Build the model.
model = Sequential([
  Dense(100, activation='relu', input_shape=(784,)),
  Dense(100, activation='relu'),
  Dense(100, activation='relu'),
  Dense(10, activation='softmax'),
])

# Compile the model.
model.compile(
  optimizer='adam',
  loss='categorical_crossentropy',
  metrics=['accuracy'],
)

# Train the model.
model.fit(
  train_images,
  to_categorical(train_labels),
  epochs=5,
  batch_size=32,
)

# Evaluate the model.
model.evaluate(
  test_images,
  to_categorical(test_labels)
)

Epoch 1/5
60000/60000 [==============================] - 10s 158us/step - loss: 0.3232 - accuracy: 0.8996
Epoch 2/5
60000/60000 [==============================] - 9s 154us/step - loss: 0.1630 - accuracy: 0.9495
Epoch 3/5
60000/60000 [==============================] - 9s 143us/step - loss: 0.1265 - accuracy: 0.9603
Epoch 4/5
60000/60000 [==============================] - 10s 169us/step - loss: 0.1063 - accuracy: 0.9665
Epoch 5/5
10000/10000 [==============================] - 2s 245us/step


[0.09791141807287931, 0.968999981880188]

Уберем один слой.

In [19]:
# Build the model.
model = Sequential([
  Dense(100, activation='relu', input_shape=(784,)),
  Dense(100, activation='relu'),
  Dense(10, activation='softmax'),
])

# Compile the model.
model.compile(
  optimizer='adam',
  loss='categorical_crossentropy',
  metrics=['accuracy'],
)

# Train the model.
model.fit(
  train_images,
  to_categorical(train_labels),
  epochs=5,
  batch_size=32,
)

# Evaluate the model.
model.evaluate(
  test_images,
  to_categorical(test_labels)
)

Epoch 1/5
60000/60000 [==============================] - 9s 145us/step - loss: 0.3273 - accuracy: 0.9008
Epoch 2/5
60000/60000 [==============================] - 8s 135us/step - loss: 0.1600 - accuracy: 0.9501
Epoch 3/5
60000/60000 [==============================] - 8s 129us/step - loss: 0.1215 - accuracy: 0.9626
Epoch 4/5
60000/60000 [==============================] - 8s 137us/step - loss: 0.1012 - accuracy: 0.9686
Epoch 5/5
10000/10000 [==============================] - 0s 30us/step


[0.09853620325699448, 0.9685999751091003]

Попробуемм добавить еще нейронов

In [20]:
# Build the model.
model = Sequential([
  Dense(150, activation='relu', input_shape=(784,)),
  Dense(150, activation='relu'),
  Dense(10, activation='softmax'),
])

# Compile the model.
model.compile(
  optimizer='adam',
  loss='categorical_crossentropy',
  metrics=['accuracy'],
)

# Train the model.
model.fit(
  train_images,
  to_categorical(train_labels),
  epochs=5,
  batch_size=32,
)

# Evaluate the model.
model.evaluate(
  test_images,
  to_categorical(test_labels)
)

Epoch 1/5
60000/60000 [==============================] - 10s 165us/step - loss: 0.2989 - accuracy: 0.9084
Epoch 2/5
60000/60000 [==============================] - 10s 162us/step - loss: 0.1461 - accuracy: 0.9551
Epoch 3/5
60000/60000 [==============================] - 9s 156us/step - loss: 0.1104 - accuracy: 0.9657
Epoch 4/5
60000/60000 [==============================] - 9s 156us/step - loss: 0.0921 - accuracy: 0.9710
Epoch 5/5
10000/10000 [==============================] - 2s 246us/step


[0.09673015461941249, 0.9713000059127808]

Попробуем добавить еще один слой и увеличить кол-во эпох.

In [21]:
# Build the model.
model = Sequential([
  Dense(150, activation='relu', input_shape=(784,)),
  Dense(150, activation='relu'),
  Dense(150, activation='relu'),
  Dense(10, activation='softmax'),
])

# Compile the model.
model.compile(
  optimizer='adam',
  loss='categorical_crossentropy',
  metrics=['accuracy'],
)

# Train the model.
model.fit(
  train_images,
  to_categorical(train_labels),
  epochs=8,
  batch_size=32,
)

# Evaluate the model.
model.evaluate(
  test_images,
  to_categorical(test_labels)
)

Epoch 1/8
60000/60000 [==============================] - 9s 152us/step - loss: 0.2959 - accuracy: 0.9073
Epoch 2/8
60000/60000 [==============================] - 10s 161us/step - loss: 0.1468 - accuracy: 0.9546
Epoch 3/8
60000/60000 [==============================] - 10s 171us/step - loss: 0.1123 - accuracy: 0.9646
Epoch 4/8
60000/60000 [==============================] - 10s 167us/step - loss: 0.0941 - accuracy: 0.9704
Epoch 5/8
60000/60000 [==============================] - 10s 161us/step - loss: 0.0822 - accuracy: 0.9736
Epoch 6/8
60000/60000 [==============================] - 9s 156us/step - loss: 0.0729 - accuracy: 0.9761
Epoch 7/8
60000/60000 [==============================] - 10s 159us/step - loss: 0.0663 - accuracy: 0.9793
Epoch 8/8
10000/10000 [==============================] - 0s 41us/step


[0.09695492423631367, 0.9726999998092651]

Результат получился лучше. По всей видимости увеличение кол-ва эпох нивелировало эффект затухающего градиента.

Попробуем еще раз увеличить все параметры.

In [22]:
# Build the model.
model = Sequential([
  Dense(180, activation='relu', input_shape=(784,)),
  Dense(180, activation='relu'),
  Dense(180, activation='relu'),
  Dense(180, activation='relu'),
  Dense(10, activation='softmax'),
])

# Compile the model.
model.compile(
  optimizer='adam',
  loss='categorical_crossentropy',
  metrics=['accuracy'],
)

# Train the model.
model.fit(
  train_images,
  to_categorical(train_labels),
  epochs=12,
  batch_size=32,
)

# Evaluate the model.
model.evaluate(
  test_images,
  to_categorical(test_labels)
)

Epoch 1/12
60000/60000 [==============================] - 12s 204us/step - loss: 0.3016 - accuracy: 0.9058
Epoch 2/12
60000/60000 [==============================] - 11s 188us/step - loss: 0.1583 - accuracy: 0.9523
Epoch 3/12
60000/60000 [==============================] - 12s 195us/step - loss: 0.1223 - accuracy: 0.9627
Epoch 4/12
60000/60000 [==============================] - 12s 194us/step - loss: 0.1023 - accuracy: 0.9682
Epoch 5/12
60000/60000 [==============================] - 11s 180us/step - loss: 0.0890 - accuracy: 0.9723
Epoch 6/12
60000/60000 [==============================] - 11s 191us/step - loss: 0.0779 - accuracy: 0.9749
Epoch 7/12
60000/60000 [==============================] - 11s 191us/step - loss: 0.0696 - accuracy: 0.9785
Epoch 8/12
60000/60000 [==============================] - 12s 198us/step - loss: 0.0668 - accuracy: 0.9788
Epoch 9/12
60000/60000 [==============================] - 11s 182us/step - loss: 0.0595 - accuracy: 0.9812
Epoch 10/12
60000/60000 [============

[0.10454919226564961, 0.9711999893188477]

Возможно 12 эпох слишком много и модель переоубчается, проверим это.

In [23]:
# Build the model.
model = Sequential([
  Dense(180, activation='relu', input_shape=(784,)),
  Dense(180, activation='relu'),
  Dense(180, activation='relu'),
  Dense(180, activation='relu'),
  Dense(10, activation='softmax'),
])

# Compile the model.
model.compile(
  optimizer='adam',
  loss='categorical_crossentropy',
  metrics=['accuracy'],
)

# Train the model.
model.fit(
  train_images,
  to_categorical(train_labels),
  epochs=10,
  batch_size=32,
)

# Evaluate the model.
model.evaluate(
  test_images,
  to_categorical(test_labels)
)

Epoch 1/10
60000/60000 [==============================] - 12s 199us/step - loss: 0.3002 - accuracy: 0.9057
Epoch 2/10
60000/60000 [==============================] - 13s 218us/step - loss: 0.1561 - accuracy: 0.9516
Epoch 3/10
60000/60000 [==============================] - 12s 205us/step - loss: 0.1241 - accuracy: 0.9614
Epoch 4/10
60000/60000 [==============================] - 12s 193us/step - loss: 0.1032 - accuracy: 0.9683
Epoch 5/10
60000/60000 [==============================] - 11s 191us/step - loss: 0.0918 - accuracy: 0.9709
Epoch 6/10
60000/60000 [==============================] - 12s 206us/step - loss: 0.0805 - accuracy: 0.9750
Epoch 7/10
60000/60000 [==============================] - 12s 195us/step - loss: 0.0726 - accuracy: 0.9769
Epoch 8/10
60000/60000 [==============================] - 12s 206us/step - loss: 0.0677 - accuracy: 0.9789
Epoch 9/10
60000/60000 [==============================] - 12s 192us/step - loss: 0.0585 - accuracy: 0.9811
Epoch 10/10
10000/10000 [============

[0.08466160992686055, 0.9771000146865845]

Результат получился чуть лучше. Попробуем добавить еще нейронов.

In [24]:
# Build the model.
model = Sequential([
  Dense(220, activation='relu', input_shape=(784,)),
  Dense(220, activation='relu'),
  Dense(220, activation='relu'),
  Dense(220, activation='relu'),
  Dense(10, activation='softmax'),
])

# Compile the model.
model.compile(
  optimizer='adam',
  loss='categorical_crossentropy',
  metrics=['accuracy'],
)

# Train the model.
model.fit(
  train_images,
  to_categorical(train_labels),
  epochs=10,
  batch_size=32,
)

# Evaluate the model.
model.evaluate(
  test_images,
  to_categorical(test_labels)
)

Epoch 1/10
60000/60000 [==============================] - 13s 213us/step - loss: 0.2964 - accuracy: 0.9071
Epoch 2/10
60000/60000 [==============================] - 12s 207us/step - loss: 0.1557 - accuracy: 0.9531
Epoch 3/10
60000/60000 [==============================] - 11s 192us/step - loss: 0.1212 - accuracy: 0.9634
Epoch 4/10
60000/60000 [==============================] - 14s 231us/step - loss: 0.1014 - accuracy: 0.9690
Epoch 5/10
60000/60000 [==============================] - 13s 225us/step - loss: 0.0895 - accuracy: 0.9724
Epoch 6/10
60000/60000 [==============================] - 13s 222us/step - loss: 0.0769 - accuracy: 0.9762
Epoch 7/10
60000/60000 [==============================] - 14s 226us/step - loss: 0.0705 - accuracy: 0.9779
Epoch 8/10
60000/60000 [==============================] - 14s 232us/step - loss: 0.0681 - accuracy: 0.9789
Epoch 9/10
60000/60000 [==============================] - 15s 242us/step - loss: 0.0599 - accuracy: 0.9805
Epoch 10/10
10000/10000 [============

[0.0924060961878713, 0.9757999777793884]

In [25]:
# Build the model.
model = Sequential([
  Dense(250, activation='relu', input_shape=(784,)),
  Dense(250, activation='relu'),
  Dense(250, activation='relu'),
  Dense(250, activation='relu'),
  Dense(250, activation='relu'),
  Dense(10, activation='softmax'),
])

# Compile the model.
model.compile(
  optimizer='adam',
  loss='categorical_crossentropy',
  metrics=['accuracy'],
)

# Train the model.
model.fit(
  train_images,
  to_categorical(train_labels),
  epochs=12,
  batch_size=32,
)

# Evaluate the model.
model.evaluate(
  test_images,
  to_categorical(test_labels)
)

Epoch 1/12
60000/60000 [==============================] - 17s 283us/step - loss: 0.3097 - accuracy: 0.9033
Epoch 2/12
60000/60000 [==============================] - 16s 267us/step - loss: 0.1631 - accuracy: 0.9513
Epoch 3/12
60000/60000 [==============================] - 17s 278us/step - loss: 0.1329 - accuracy: 0.9606
Epoch 4/12
60000/60000 [==============================] - 17s 291us/step - loss: 0.1086 - accuracy: 0.9674
Epoch 5/12
60000/60000 [==============================] - 17s 290us/step - loss: 0.0969 - accuracy: 0.9714
Epoch 6/12
60000/60000 [==============================] - 15s 258us/step - loss: 0.0848 - accuracy: 0.9743
Epoch 7/12
60000/60000 [==============================] - 15s 250us/step - loss: 0.0791 - accuracy: 0.9761
Epoch 8/12
60000/60000 [==============================] - 16s 269us/step - loss: 0.0702 - accuracy: 0.9791
Epoch 9/12
60000/60000 [==============================] - 15s 251us/step - loss: 0.0674 - accuracy: 0.9800
Epoch 10/12
60000/60000 [============

[0.11230551293905337, 0.9758999943733215]

In [26]:
# Build the model.
model = Sequential([
  Dense(180, activation='relu', input_shape=(784,)),
  Dense(180, activation='relu'),
  Dense(180, activation='relu'),
  Dense(180, activation='relu'),
  Dense(180, activation='relu'),
  Dense(10, activation='softmax'),
])

# Compile the model.
model.compile(
  optimizer='adam',
  loss='categorical_crossentropy',
  metrics=['accuracy'],
)

# Train the model.
model.fit(
  train_images,
  to_categorical(train_labels),
  epochs=12,
  batch_size=32,
)

# Evaluate the model.
model.evaluate(
  test_images,
  to_categorical(test_labels)
)

Epoch 1/12
60000/60000 [==============================] - 14s 230us/step - loss: 0.3126 - accuracy: 0.9013
Epoch 2/12
60000/60000 [==============================] - 15s 247us/step - loss: 0.1657 - accuracy: 0.9504
Epoch 3/12
60000/60000 [==============================] - 14s 238us/step - loss: 0.1326 - accuracy: 0.9599
Epoch 4/12
60000/60000 [==============================] - 15s 253us/step - loss: 0.1100 - accuracy: 0.9669
Epoch 5/12
60000/60000 [==============================] - 15s 247us/step - loss: 0.0970 - accuracy: 0.9703
Epoch 6/12
60000/60000 [==============================] - 15s 252us/step - loss: 0.0846 - accuracy: 0.9742
Epoch 7/12
60000/60000 [==============================] - 14s 230us/step - loss: 0.0789 - accuracy: 0.9762
Epoch 8/12
60000/60000 [==============================] - 14s 237us/step - loss: 0.0707 - accuracy: 0.9783
Epoch 9/12
60000/60000 [==============================] - 14s 239us/step - loss: 0.0670 - accuracy: 0.9797
Epoch 10/12
60000/60000 [============

[0.08670836529053703, 0.9775000214576721]

In [29]:
# Build the model.
model = Sequential([
  Dense(180, activation='relu', input_shape=(784,)),
  Dense(180, activation='relu'),
  Dense(180, activation='relu'),
  Dense(180, activation='relu'),
  Dense(180, activation='relu'),
  Dense(10, activation='softmax'),
])

# Compile the model.
model.compile(
  optimizer='adam',
  loss='categorical_crossentropy',
  metrics=['accuracy'],
)

# Train the model.
model.fit(
  train_images,
  to_categorical(train_labels),
  epochs=11,
  batch_size=32,
)

# Evaluate the model.
model.evaluate(
  test_images,
  to_categorical(test_labels)
)

Epoch 1/11
60000/60000 [==============================] - 16s 264us/step - loss: 0.3097 - accuracy: 0.9025
Epoch 2/11
60000/60000 [==============================] - 16s 265us/step - loss: 0.1637 - accuracy: 0.9502
Epoch 3/11
60000/60000 [==============================] - 15s 246us/step - loss: 0.1328 - accuracy: 0.9594
Epoch 4/11
60000/60000 [==============================] - 15s 247us/step - loss: 0.1132 - accuracy: 0.9659
Epoch 5/11
60000/60000 [==============================] - 15s 255us/step - loss: 0.0956 - accuracy: 0.9713
Epoch 6/11
60000/60000 [==============================] - 14s 241us/step - loss: 0.0905 - accuracy: 0.9726
Epoch 7/11
60000/60000 [==============================] - 15s 257us/step - loss: 0.0765 - accuracy: 0.9766
Epoch 8/11
60000/60000 [==============================] - 16s 262us/step - loss: 0.0691 - accuracy: 0.9788
Epoch 9/11
60000/60000 [==============================] - 15s 257us/step - loss: 0.0666 - accuracy: 0.9797
Epoch 10/11
60000/60000 [============

[0.10209386850666488, 0.9718999862670898]

In [31]:
# Build the model.
model = Sequential([
  Dense(180, activation='relu', input_shape=(784,)),
  Dense(180, activation='relu'),
  Dense(180, activation='relu'),
  Dense(180, activation='relu'),
  Dense(180, activation='relu'),
  Dense(180, activation='relu'),
  Dense(10, activation='softmax'),
])

# Compile the model.
model.compile(
  optimizer='adam',
  loss='categorical_crossentropy',
  metrics=['accuracy'],
)

# Train the model.
model.fit(
  train_images,
  to_categorical(train_labels),
  epochs=14,
  batch_size=32,
)

# Evaluate the model.
model.evaluate(
  test_images,
  to_categorical(test_labels)
)

Epoch 1/14
60000/60000 [==============================] - 16s 275us/step - loss: 0.3295 - accuracy: 0.8970
Epoch 2/14
60000/60000 [==============================] - 16s 269us/step - loss: 0.1793 - accuracy: 0.9471
Epoch 3/14
60000/60000 [==============================] - 16s 267us/step - loss: 0.1422 - accuracy: 0.9582
Epoch 4/14
60000/60000 [==============================] - 15s 245us/step - loss: 0.1184 - accuracy: 0.9646
Epoch 5/14
60000/60000 [==============================] - 14s 230us/step - loss: 0.1058 - accuracy: 0.9692
Epoch 6/14
60000/60000 [==============================] - 15s 248us/step - loss: 0.0925 - accuracy: 0.9728
Epoch 7/14
60000/60000 [==============================] - 15s 256us/step - loss: 0.0842 - accuracy: 0.9750
Epoch 8/14
60000/60000 [==============================] - 15s 246us/step - loss: 0.0759 - accuracy: 0.9774
Epoch 9/14
60000/60000 [==============================] - 15s 247us/step - loss: 0.0716 - accuracy: 0.9793
Epoch 10/14
60000/60000 [============

[0.08889369516318621, 0.9796000123023987]

В итоге лучший результат с точностью 0.9796 мы получили на нейросети с 7-ю слоями по 180 нейронов в каждом, кроме последнего и с 14 эпохами обучения.

Возможно что некоторые отличия результатов это результат разных начальных весов. Увидели тенденцию на увеличение точности с ростом кол-ва слоев и нейронов в них. Модель с большим количеством слоев сложнее обучать. Итого разница между лучшей и худшей моделью оказалась чуть больше 2%.